src/config/index.ts

```diff
@@ -1,12 +1,20 @@
 import { Config } from './../types';
+import preprocessConfig from './preprocess-config';
 import processConfig from './process-config';
 import readConfig from './read-config';
 import validateConfig from './validate-config';
 
 export const getConfig = (configPath?: string): Config => {
     const foundConfig = readConfig(configPath);
-    validateConfig(foundConfig);
-    const processedConfig = processConfig(foundConfig);
+    const preprocessedConfig = {
+        ...foundConfig,
+        config: foundConfig && foundConfig.config ? preprocessConfig(foundConfig) : undefined,
+    };
+    if (validateConfig(preprocessedConfig)) {
```

Comment:
Could you, please, update `validateConfig` to return `void` and remove this condition as well as `throw Error('Invalid config');` line below?

Reply:
The reason why I changed the signature of `validateConfig` was to use [User-defined type guards](https://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards) to infer the type of `preprocessedConfig` to `ValidatedConfigFile` without the use of type casting.

Unfortunately it only works if the type guard function is a boolean and there's no way of signalling that a function throws an error instead.

I can revert this and just typecast `preprocessedConfig` instead if you think that's cleaner:

```js
validateConfig(preprocessedConfig);
const processedConfig = processConfig(preprocessedConfig as ValidatedConfigFile);
return processedConfig;
```

I agree that the dead code to satisfy the type system is not ideal.

Reply:
Thanks for the explanation.
Yes, please, let's revert it. We might start using type guards in future, but right now they make you write code in an unusual way.